src/keys.rs

```diff
@@ -1,70 +1,193 @@
 //! Key structures for Orchard.
 
-use crate::address::Address;
+use std::convert::TryInto;
+use std::mem;
+
+use aes::Aes256;
+use fpe::ff1::{BinaryNumeralString, FF1};
+use group::GroupEncoding;
+use halo2::{arithmetic::FieldExt, pasta::pallas};
+use subtle::CtOption;
+
+use crate::{
+    address::Address,
+    primitives::redpallas::{self, SpendAuth},
+    spec::{
+        commit_ivk, diversify_hash, extract_p, ka_orchard, prf_expand, prf_expand_vec, to_base,
+        to_scalar,
+    },
+};
 
 /// A spending key, from which all key material is derived.
 ///
-/// TODO: In Sapling we never actually used this, instead deriving everything via ZIP 32,
-/// so that we could maintain Bitcoin-like HD keys with properties like non-hardened
-/// derivation. If we decide that we don't actually require non-hardened derivation, then
-/// we could greatly simplify the HD structure and use this struct directly.
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][§4.2.3].
+///
+/// [§4.2.3]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
```

Comment:
Is there a danger of these numbers changing? If so, maybe using `[spec]` or maybe `[orchardkeycomponents]` will be less likely to become inaccurate

Reply:
If the numbers become inaccurate, they'd be inaccurate in the link text as well as the underlying link handle, and if we're updating the former then the latter would be handled at the same time. That being said, I'm fine with having the link handles follow the URL anchors.